crates/noirc_evaluator/src/ssa/conditional.rs

```diff
@@ -156,92 +174,114 @@ impl DecisionTree {
     }
 
     pub fn make_decision_tree(&mut self, ctx: &mut SsaContext, entry_block: BlockId) {
-        let mut join_to_process = Vec::new();
-        let mut join_processed = Vec::new();
-        let mut stack = StackFrame::new(entry_block);
-        self.decision_tree(ctx, self.root, &mut join_to_process, &mut join_processed, &mut stack);
+        let mut builder = TreeBuilder::new();
+        builder.stack.block = entry_block;
+        ctx[entry_block].assumption = self.root;
+        self.decision_tree(ctx, entry_block, &mut builder);
     }
 
-    pub fn decision_tree(
+    //Returns a boolean to indicate if we should process the children (true) of not (false) of the block
+    fn process_block(
         &mut self,
         ctx: &mut SsaContext,
-        current_assumption: AssumptionId,
-        join_to_process: &mut Vec<BlockId>,
-        join_processed: &mut Vec<BlockId>,
-        stack: &mut StackFrame,
-    ) {
-        let assumption = &self[current_assumption];
-        let block = &ctx[stack.block];
-        let mut block_assumption = current_assumption;
+        current: BlockId,
+        data: &mut TreeBuilder,
+    ) -> bool {
+        data.stack.block = current;
+        let mut block_assumption = ctx[current].assumption;
+        let assumption = &self[block_assumption];
+
+        let current_block = &ctx[current];
         let mut if_assumption = None;
         let mut parent = AssumptionId::dummy();
         let mut sibling = true;
-        if join_processed.contains(&stack.block) {
-            return;
+        if let Some(pos) = data.join_processed.iter().position(|value| *value == current) {
+            data.join_processed.remove(pos);
```

Comment:
Should `join_processed` be a set instead? It would be more efficient to remove them if so.

Reply:
I used swap_remove() instead. I don't expect the processed list to grow a lot because it is cleaned here so I believe the insert should be cheap (but I am not 100% sure)